src/bootstrap/src/ferrocene/code_coverage.rs

```diff
@@ -1,90 +1,73 @@
-use std::path::Path;
+use std::path::PathBuf;
 
-use crate::builder::{Builder, RunConfig, ShouldRun, Step};
-use crate::core::build_steps::tool::SourceType;
+use crate::BootstrapCommand;
+use crate::builder::Builder;
+use crate::core::build_steps::llvm::Llvm;
+use crate::core::builder::Cargo;
 use crate::core::config::TargetSelection;
-use crate::{BootstrapCommand, Kind, Mode};
 
-#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub(crate) struct ProfilerBuiltinsNoCore {
-    pub(crate) target: TargetSelection,
+pub(crate) fn instrument_coverage(cargo: &mut Cargo) {
+    cargo.rustflag("-Cinstrument-coverage");
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub(crate) struct InstrumentationCoverageFlags {
-    rustflags: Vec<String>,
+pub(crate) struct GatherCoverage<'a> {
+    builder: &'a Builder<'a>,
+    llvm_bin_dir: PathBuf,
+    profraw_dir: PathBuf,
+    profdata_file: PathBuf,
 }
 
-impl InstrumentationCoverageFlags {
-    pub fn flags(&self) -> &[String] {
-        &self.rustflags
-    }
-}
-
-impl Step for ProfilerBuiltinsNoCore {
-    type Output = InstrumentationCoverageFlags;
-    const ONLY_HOSTS: bool = true;
-    const DEFAULT: bool = true;
-
-    fn run(self, builder: &Builder<'_>) -> Self::Output {
-        builder.info("Building profiler_builtins with no_core");
-
-        let target = self.target;
-        let compiler = builder.compiler(1, target);
-
-        // x.py test --coverage
-        // [ProfilerBuiltinNoCore step] Builds profiler_builtins with no core -> profiler_builtins_no_core.rlib
-        // [Core Tests step] Compile tests of core with -Cinsturment-coverage -> link against profiler_builtins_no_core.rlib
-
-        let mut cargo = builder.cargo(compiler, Mode::Std, SourceType::InTree, target, Kind::Build);
-
-        cargo.current_dir(Path::new("library/profiler_builtins"));
-
-        let profiler_builtins_no_core_dir = "profiler_builtins_no_core";
-        let target_dir =
-            builder.cargo_out(compiler, Mode::Std, target).join(profiler_builtins_no_core_dir);
+impl<'a> GatherCoverage<'a> {
+    pub(crate) fn new(
+        builder: &'a Builder<'a>,
+        cargo: &mut Cargo,
+        target: TargetSelection,
+        name: &str,
```

Comment:
What's the purpose of the `name` if it is always `"library"`?

Reply:
Maybe you can add a doc-comment how you expect it to be used?

Reply:
I will replace this with a different mechanism in a separate PR (for different reasons).